### PR TITLE
Generate Dart enum variants in camelCase.

### LINKED ISSUE #1110: Dart enums should have their variant names in Dart style.
**Is your feature request related to a problem? Please describe.**
I am wanting to reduce code reuse in my project. I realized that the comments from Rust are transferred to Dart.  Previously, I created my own enum to document it. However, what is stopping me from using the generated enum is that the enum follows Rust's conventions.

**Describe the solution you'd like**
`MyDartEnum.VariantOne` -> `MyDartEnum.variantOne`

**Describe alternatives you've considered**
I made my own enum that is a 1 to 1 copy of the generated Dart enum to use in the public API of my project.

**Additional context**
I did some testing and modified [this line](https://github.com/fzyzcjy/flutter_rust_bridge/blob/617d52ed1c6f9d92f0d485fdafea1e034edf34ef/frb_codegen/src/generator/dart/ty_enum.rs#L241) to use `dart_style()` and it did change the generated enum's values. However, I couldn't get the code gen to work properly in the pure_dart example so I couldn't continue testing.

### PR DESCRIPTION
**Pull request by erikas-taroza**
## Changes
Fixes #1110 

The generated Dart enums use proper casing.
`MyEnum.VariantOne` -> `MyEnum.variantOne`

Todo List:
- [x] Generate enum with proper variant names.
- [x] Handle `#[frb(default = "Weekdays.Sunday")]`
- [x] Handle enum variants that are Dart keywords.

## Checklist

- [x] An issue to be fixed by this PR is listed above.
- [x] New tests are added to ensure new features are working. End-to-end tests are usually in the `./frb_example/pure_dart` example, more specifically, `rust/src/api.rs` and `dart/lib/main.dart`.
- [x] The code generator is run and the code is formatted (via `just precommit`).
- [x] If this PR adds/changes features, documentations (in the `./book` folder) are updated.
- [x] CI is passing.

## Remark for PR creator

- Justfile is a task runner for the command line interface that allows you to run shell commands from a file named `justfile` in your project directory. You can use Justfile after [installing it](https://github.com/casey/just). Note that commands written in `justfile` of this repository are expected to be run in `bash`, not `cmd` or `powershell`. Running `just ...` commands in `cmd` or `powershell` will produce errors as the syntax is not compatible. On Windows, you can use `git bash` if you have Git installed.
- If fzyzcjy does not reply for a few days, maybe he just did not see it, so please ping him.

### CODE REVIEW

**Comment by erikas-taroza** [comment]
Hello @fzyzcjy 

I have a question about my changes in 28ec345. It addresses the `#[frb(default = "Weekdays.Sunday")]` macro in `api.rs`. Since this macro is used in Rust, I think it would make sense to keep the casing in PascalCase for the user. Instead of having them modify the macro, the codegen corrects the casing for Dart. Do the changes in my commit have any side effects? I ran the codegen on the `pure_dart` example and it seems to work as expected. (I don't know the code base.)

Also, the codegen fails for `MyEnum` which has variants `False` and `True`. These are keywords in Dart when lowercased. How do you think this should be handled? I am thinking about comparing to a list of Dart keywords and deciding if PascalCase is needed. Or the user should not have their variants named after keywords (and let the generated code cause an error).

**Comment by fzyzcjy** [comment]
Good job!

> Do the changes in my commit have any side effects?

Just ensure the CI passes ;) That's one of ways people check whether a code breaks sth unexpectedly on a large codebase

>  These are keywords in Dart when lowercased. How do you think this should be handled?

Maybe we can make it `true_` and `false_`

**Comment by fzyzcjy** [comment]
In addition, maybe we should add a flag to the code generator to explicitly enable this feature. Because otherwise, this is a breaking change to existing users.

**Comment by erikas-taroza** [comment]
> Maybe we can make it true_ and false_

The issue is not really with naming but how to detect if the variant in lowercase is a Dart keyword. Maybe a method in IrIdent? Something called `is_dart_keyword` that would take a string, camelCase it, and be checked in a match statement.

> maybe we should add a flag to the code generator to explicitly enable this feature.

Good idea. I'm going to bed so I'll add this when I wake up.

**Comment by fzyzcjy** [comment]
Sure, take your time!

> how to detect if the variant in lowercase is a Dart keyword.

Not looked at your PR code detail yet, but I guess maybe we should do something like, "when you are going to output that (lowercased) variant, check your lowercased variant to see whether it is a keyword"

**Comment by erikas-taroza** [comment]
> Not looked at your PR code detail yet, but I guess maybe we should do something like, "when you are going to output that (lowercased) variant, check your lowercased variant to see whether it is a keyword"

Like a piece of documentation?



> In addition, maybe we should add a flag to the code generator to explicitly enable this feature.

I have added a flag called `--dart-enums-style`.

File: frb_codegen/Cargo.toml [hunk 1]
```diff
@@ -56,3 +56,4 @@ x86_64-pc-windows-msvc = {pkg-fmt = "zip"}
 chrono = []
 default = ["chrono"]
 uuid = []
+dart-style-enums = []
```
Comment (fzyzcjy) [inline frb_codegen/Cargo.toml]: Well, I mean one (command line) flag, e.g. in `Opts`/`RawOpts` etc

Comment (erikas-taroza) [reply frb_codegen/Cargo.toml]: Yeah I removed it. Side product of just waking up :D

**Comment by erikas-taroza** [comment]
In db07d8e, I have added a function in `utils.rs` to check if the generated variant name is a keyword. If it is a keyword, the casing will remain in PascalCase.

**Comment by erikas-taroza** [comment]
Not sure why the CI is failing. I updated everything in `main.dart` :thinking: 

```
frb_example/pure_dart/dart/lib/main.dart:213:32: Error: Member not found: 'friday'.
    expect(r.weekday, Weekdays.friday);
                               ^^^^^^
frb_example/pure_dart/dart/lib/main.dart:374:67: Error: Member not found: 'tuesday'.
    expect(await api.handleReturnEnum(input: "Tuesday"), Weekdays.tuesday);
                                                                  ^^^^^^^
frb_example/pure_dart/dart/lib/main.dart:[379](https://github.com/fzyzcjy/flutter_rust_bridge/actions/runs/4370396073/jobs/7645265735#step:7:380):60: Error: Member not found: 'saturday'.
    expect(await api.handleEnumParameter(weekday: Weekdays.saturday), Weekdays.saturday);
                                                           ^^^^^^^^
frb_example/pure_dart/dart/lib/main.dart:379:80: Error: Member not found: 'saturday'.
    expect(await api.handleEnumParameter(weekday: Weekdays.saturday), Weekdays.saturday);
                                                                               ^^^^^^^^
frb_example/pure_dart/dart/lib/main.dart:399:66: Error: Member not found: 'monday'.
      await api.handleEnumStruct(val: KitchenSink_Enums(Weekdays.monday)),
                                                                 ^^^^^^
frb_example/pure_dart/dart/lib/main.dart:400:34: Error: Member not found: 'tuesday'.
      KitchenSink_Enums(Weekdays.tuesday),
                                 ^^^^^^^
frb_example/pure_dart/dart/lib/main.dart:433:43: Error: Member not found: 'standalone'.
    expect(settings.mode, ApplicationMode.standalone);
                                          ^^^^^^^^^^
frb_example/pure_dart/dart/lib/main.dart:443:39: Error: Member not found: 'embedded'.
                mode: ApplicationMode.embedded,
                                      ^^^^^^^^
frb_example/pure_dart/dart/lib/main.dart:1252:83: Error: Member not found: 'friday'.
  return MyNestedStruct(treeNode: _createMyTreeNode(arrLen: 5), weekday: Weekdays.friday);
                                                                                  ^^^^^^
```

**Comment by fzyzcjy** [comment]
Maybe run `just precommit` in your local computer to see what is wrong. Then, if you want to speed up, see which sub-part in the `justfile` is going wrong and run that (e.g. `just test_whatever`')

**Comment by erikas-taroza** [comment]
> Maybe run `just precommit` in your local computer to see what is wrong. Then, if you want to speed up, see which sub-part in the `justfile` is going wrong and run that (e.g. `just test_whatever`')

Nothing changes except for some formatting in Rust.

**Comment by fzyzcjy** [comment]
Btw you may modify the CI yaml file for debugging. For example, `cat whatever_file.dart` and see its content.

**Comment by erikas-taroza** [comment]
> Btw you may modify the CI yaml file for debugging. For example, `cat whatever_file.dart` and see its content.

The `frb_example/pure_dart/dart/lib/main.dart` file is the updated one but the CI is still failing. In `Dart test with Valgrind`, running the command that caused the error (`dart compile exe lib/main.dart -o main.exe`) works on my machine as expected.

I have no clue what is happening here.

**Comment by erikas-taroza** [comment]
> > main.dart file is the updated
> 
> what about the bridge_generated.dart etc

:facepalm: Looking at the wrong file. Lets see...

**Comment by erikas-taroza** [comment]
`bridge_definitions.dart` contains the enum with the correct names. I'm going to bed now. I will be travelling tomorrow so I may not be able to work on this.

**Comment by fzyzcjy** [comment]
Take your time!

**Comment by erikas-taroza** [comment]
Changing `pure_dart/dart/lib/main.dart` to use the old variant names passes the 3 failing tests but fails other tests (like `Dart linter`)  because those are the old variant names.

**Comment by erikas-taroza** [comment]
Fixed the CI! Just need to let it finish to make sure.

I cloned my fork again and when I opened it in VSCode, it was running codegen. I saw that the` bridge_definitions.dart` was generated using the Rust style variant names. It turns out that the `build.rs` in the `pure_dart` example was running the codegen without the new flag and thus causing the error. What a sneaky bug!

**Comment by fzyzcjy** [comment]
Good job! I get confused: Why `just precommit` does not figure out this bug? Then it seems like a bug in infra as well.

**Comment by erikas-taroza** [comment]
> Good job! I get a bit confused: Why `just precommit` does not figure out this bug? Then it seems like a bug in infra as well - `just precommit` should at least run codegen (implicitly) using build.rs so this should be spotted at the beginning.

I'm on Fedora so maybe I missed a command? I have to copy and paste the commands, it doesn't work OOB. I am too lazy to configure it.

**Comment by fzyzcjy** [comment]
I see. Will review this PR in probably today :)

**Comment by erikas-taroza** [comment]
> I see. Will review this PR in probably today :)

Sounds good :)

File: frb_codegen/src/generator/dart/ty_enum.rs [hunk 2]
```diff
@@ -235,10 +236,20 @@ impl TypeDartGeneratorTrait for TypeEnumRefGenerator<'_> {
                 .variants()
                 .iter()
                 .map(|variant| {
+                    let variant_name = if self.context.config.dart_enums_style {
+                        if crate::utils::check_for_keywords(&[variant.name.dart_style()]).is_err() {
+                            variant.name.dart_style().to_case(Case::Pascal)
+                        } else {
+                            variant.name.dart_style().to_case(Case::Camel)
+                        }
```
Comment (fzyzcjy) [inline frb_codegen/src/generator/dart/ty_enum.rs:244]: uber-nit: maybe sth like:

crate::utils::auto_rename_avoid_dart_keyword_collision(variant.name.dart_style().to_case(Case::Camel))

and

fn auto_rename_avoid_dart_keyword_collision(s: &str) {
  if check_for_keywords { s + "_" } else { s}
}

(pseudocode and pseudo function name only ;) )

File: frb_codegen/src/ir/field.rs [hunk 3]
```diff
@@ -18,15 +20,32 @@ impl IrField {
...
+                            let mut split = value.split('.');
+                            let enum_name = split.next().unwrap();
+
+                            let variant_name = split.next().unwrap().to_string();
+                            let variant_name =
+                                if crate::utils::check_for_keywords(&[variant_name.to_case(Case::Camel)]).is_err() {
+                                    variant_name.to_case(Case::Pascal)
+                                } else {
+                                    variant_name.to_case(Case::Camel)
+                                };
```
Comment (fzyzcjy) [inline frb_codegen/src/ir/field.rs:43]: uber-nit: If we use the approach above, here we can reuse that avoid-collision function

File: frb_codegen/src/ir/field.rs [hunk 4]
```diff
@@ -18,15 +20,32 @@ impl IrField {
...
+    pub fn field_default(&self, freezed: bool, config: Option<&Opts>) -> String {
         self.default
             .as_ref()
             .map(|r#default| {
                 let r#default = match r#default {
                     DefaultValues::Str(lit)
                         if !matches!(&self.ty, IrType::Delegate(IrTypeDelegate::String)) =>
                     {
-                        lit.value().into()
+                        // Convert the default value to Dart style.
```
Comment (fzyzcjy) [inline frb_codegen/src/ir/field.rs:31]: uber-nit: what about extracting this part into a new (private) function

**Review by fzyzcjy — COMMENTED** [review]
Ready to merge, only a few nits :)

**Comment by erikas-taroza** [comment]
@fzyzcjy  Fixed :)

**Review by fzyzcjy — APPROVED** [review]

**Comment by fzyzcjy** [comment]
@all-contributors please add @erikas-taroza for code

**Pull request merged**